src/cpp/core/include/core/Database.hpp

```diff
@@ -0,0 +1,282 @@
+{
+public:
+   virtual ~Connection() {}
+
+   Query query(const std::string& sqlStatement) override;
+
+   Error execute(Query& query,
+                 Rowset& rowset) override;
+
+   Error execute(Query& query,
+                 bool* pDataReturned = nullptr) override;
+
+   Error executeStr(const std::string& queryStr) override;
+
+   std::string driverName() const override;
+
+   soci::session& session() override { return session_; }
+
+private:
+   friend class ConnectVisitor;
+   friend class Transaction;
+
+   // private constructor - use global connect function
+   Connection(const soci::backend_factory& factory,
+              const std::string& connectionStr);
+
+   soci::session session_;
+};
+
+class PooledConnection : public IConnection
+{
+public:
+   virtual ~PooledConnection();
+
+   Query query(const std::string& sqlStatement) override;
+
+   Error execute(Query& query,
+                 Rowset& rowset) override;
+
+   Error execute(Query& query,
+                 bool* pDataReturned = nullptr) override;
+
+   Error executeStr(const std::string& queryStr) override;
+
+   std::string driverName() const override;
+
+   soci::session& session() override { return connection_->session(); }
+
+private:
+   friend class ConnectionPool;
+
+   // private constructor - get PooledConnection from ConnectionPool
+   PooledConnection(const boost::shared_ptr<ConnectionPool>& pool,
+                    const boost::shared_ptr<Connection>& connection);
+
+   boost::shared_ptr<ConnectionPool> pool_;
+   boost::shared_ptr<Connection> connection_;
+};
+
+class ConnectionPool : public boost::enable_shared_from_this<ConnectionPool>
+{
+public:
+   boost::shared_ptr<IConnection> getConnection();
+
+private:
+   friend class PooledConnection;
+   friend Error createConnectionPool(size_t poolSize,
+                                     const ConnectionOptions& options,
+                                     boost::shared_ptr<ConnectionPool>* pPool);
+
+   void returnConnection(const boost::shared_ptr<Connection>& connection);
+
+   thread::ThreadsafeQueue<boost::shared_ptr<Connection>> connections_;
+};
+
+class Transaction
+{
+public:
+   Transaction(const boost::shared_ptr<IConnection>& connection);
+
+   void commit();
+   void rollback();
+
+   // when this class goes out of scope, the transaction
+   // is automatically aborted if not previously committed
+
+private:
+   boost::shared_ptr<IConnection> connection_;
+   soci::transaction transaction_;
+};
+
+static constexpr const char* SQLITE_DRIVER = "sqlite3";
+static constexpr const char* POSTGRESQL_DRIVER = "postgresql";
+
+class SchemaUpdater
+{
+public:
+   SchemaUpdater(const boost::shared_ptr<IConnection>& connection,
+                 const FilePath& migrationsPath);
+
+   // updates the database schema to the highest version
+   Error update();
+
+   // updates the database schema to the specified version
+   Error updateToVersion(const std::string& maxVersion);
```

Comment:
Both - it will update to no higher than the specified version, and if that version exists, that's what it will be updated to.

But that break condition was something I forgot to add to the code, so will update with that in there.

Reply:
Could you update the comment to reflect that as well?